components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +116,91 @@ impl Dicts {
         }
     }
 
-    fn save_key_dict(&mut self, master_key: &dyn Backend) -> Result<()> {
+    fn save_key_dict(&self, master_key: &dyn Backend) -> Result<()> {
+        let _lock = self.key_lock.lock().unwrap();
         let file = EncryptedFile::new(&self.base, KEY_DICT_NAME);
+        let mut key_dict = self.key_dict.lock().unwrap();
         if !master_key.is_secure() {
-            for value in self.key_dict.keys.values_mut() {
+            for value in key_dict.keys.values_mut() {
                 value.was_exposed = true
             }
         }
-        let key_bytes = self.key_dict.write_to_bytes()?;
+        key_dict.current_key_id = self.current_key_id.load(Ordering::SeqCst);
+        let keys_len = key_dict.keys.len() as _;
+        let key_bytes = key_dict.write_to_bytes()?;
+        drop(key_dict);
         file.write(&key_bytes, master_key)?;
 
         ENCRYPTION_FILE_SIZE_GAUGE
             .with_label_values(&["key_dictionary"])
             .set(key_bytes.len() as _);
-        ENCRYPTION_DATA_KEY_GAUGE.set(self.key_dict.keys.len() as _);
+        ENCRYPTION_DATA_KEY_GAUGE.set(keys_len);
 
         Ok(())
     }
 
     fn save_file_dict(&self) -> Result<()> {
+        let _lock = self.file_lock.lock().unwrap();
+        let file_dict = self.file_dict.lock().unwrap();
         let file = EncryptedFile::new(&self.base, FILE_DICT_NAME);
-        let file_bytes = self.file_dict.write_to_bytes()?;
+        let file_bytes = file_dict.write_to_bytes()?;
+        let file_num = file_dict.files.len() as _;
+        drop(file_dict);
```

Comment:
`drop` can be avoid by scoping. and `EncryptedFile::new` can be moved to out of mutex. Same for all occurrence of `drop`. 
```
let file = EncryptedFile::new(&self.base, FILE_DICT_NAME);
let file_bytes, file_num = {
    let file_dict = self.file_dict.lock().unwrap();
    let file_bytes = file_dict.write_to_bytes()?
    let file_num = file_dict.files.len() as _;
    (file_bytes, file_num)
}
```